plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataBuilder.java

```diff
@@ -0,0 +1,356 @@
+      // release version does not have version-level metadata
+      log.debug("Not a snapshot or nothing to generate: {}:{}:{}", getGroupId(), getArtifactId(), getBaseVersion());
+      return null;
+    }
+    final List<Snapshot> snapshots = Lists.newArrayList();
+    for (VersionCoordinates versionCoordinates : latestVersionCoordinatesMap.values()) {
+      final Coordinates coordinates = versionCoordinates.coordinates;
+      final Snapshot snapshotVersion = Maven2Metadata.newSnapshot(
+          new DateTime(coordinates.getTimestamp()),
+          coordinates.getExtension(),
+          coordinates.getClassifier(),
+          coordinates.getVersion()
+      );
+      snapshots.add(snapshotVersion);
+    }
+    return Maven2Metadata.newVersionLevel(
+        DateTime.now(),
+        getGroupId(),
+        getArtifactId(),
+        getBaseVersion(),
+        latestVersionCoordinates.coordinates.getTimestamp(),
+        latestVersionCoordinates.coordinates.getBuildNumber(),
+        snapshots
+    );
+  }
+
+  public void addArtifactVersion(final MavenPath mavenPath) {
+    checkNotNull(mavenPath, "mavenPath");
+    if (mavenPath.isSubordinate() || mavenPath.getCoordinates() == null) {
+      return;
+    }
+    checkState(Objects.equals(getGroupId(), mavenPath.getCoordinates().getGroupId()));
+    checkState(Objects.equals(getArtifactId(), mavenPath.getCoordinates().getArtifactId()));
+    checkState(Objects.equals(getBaseVersion(), mavenPath.getCoordinates().getBaseVersion()));
+
+    log.debug("Discovered {}:{}:{}:{}:{}",
+        mavenPath.getCoordinates().getGroupId(),
+        mavenPath.getCoordinates().getArtifactId(),
+        mavenPath.getCoordinates().getVersion(),
+        mavenPath.getCoordinates().getClassifier(),
+        mavenPath.getCoordinates().getExtension());
+
+    addBaseVersion(mavenPath.getCoordinates().getBaseVersion());
+
+    if (!mavenPath.getCoordinates().isSnapshot()) {
+      return;
+    }
+    if (Objects.equals(mavenPath.getCoordinates().getBaseVersion(), mavenPath.getCoordinates().getVersion())) {
+      log.warn("Non-timestamped snapshot, ignoring it: {}", mavenPath);
+      return;
+    }
+
+    final String key = key(mavenPath.getCoordinates());
+    final Version version = parseVersion(mavenPath.getCoordinates().getVersion());
+    final VersionCoordinates versionCoordinates = new VersionCoordinates(version, mavenPath.getCoordinates());
+
+    // maintain latestVersionCoordinates
+    if (latestVersionCoordinates == null || latestVersionCoordinates.version.compareTo(version) < 0) {
+      latestVersionCoordinates = versionCoordinates;
+    }
+
+    // maintain latestVersionCoordinatesMap
+    if (!latestVersionCoordinatesMap.containsKey(key)) {
+      latestVersionCoordinatesMap.put(key, versionCoordinates);
+    }
+    else {
+      // add if contained version is less than version
+      final VersionCoordinates other = latestVersionCoordinatesMap.get(key);
```

Comment:
could just get then include a null check before comparing, which would avoid the need to first call `containsKey`